### PR TITLE
Convert product-elements/tag-list to TypeScript

### LINKED ISSUE #7098: Convert `product-elements/tag-list` to TypeScript
Folder: https://github.com/woocommerce/woocommerce-blocks/tree/trunk/assets/js/atomic/blocks/product-elements/tag-list

### PR DESCRIPTION
**Pull request by nielslange**
Fixes #7098

Currently, the `product-elements/tag-list` is still written in JS and is using [prop-types](https://www.npmjs.com/package/prop-types) for type checking. This PR aims to convert `product-elements/tag-list` to TypeScript.

#### User Facing Testing

1. Create a test page and add the `Single Product block` to it.
2. Verify that the `Product Tag List inner block` is still visible.

* [x] Do not include in the Testing Notes

### CODE REVIEW

File: assets/js/atomic/blocks/product-elements/tag-list/constants.tsx [hunk 1]
```diff
@@ -4,14 +4,14 @@
...
+export const BLOCK_TITLE: string = __(
 	'Product Tag List',
 	'woo-gutenberg-products-block'
 );
-export const BLOCK_ICON = (
+export const BLOCK_ICON: JSX.Element = (
 	<Icon icon={ tag } className="wc-block-editor-components-block-icon" />
 );
-export const BLOCK_DESCRIPTION = __(
+export const BLOCK_DESCRIPTION: string = __(
```
Comment (dinhtungdu) [inline assets/js/atomic/blocks/product-elements/tag-list/constants.tsx:14]: TS can guess the type of these variables. I think we should rely on Type Inference when possible, what do you think?

**Review by dinhtungdu — COMMENTED** [review]
The code looks good and the block works as expected in my test. Left a minor comment about Type Inference but otherwise, this is LGTM!

File: assets/js/atomic/blocks/product-elements/tag-list/constants.tsx (hunk 1 above)
Comment (nielslange) [reply assets/js/atomic/blocks/product-elements/tag-list/constants.tsx:14]: Thanks for your review, @dinhtungdu. 

> TS can guess the type of these variables. I think we should rely on Type Inference when possible, what do you think?

I think slightly different on that. I believe we should always declare the type, even though TS can guess the type. In the example you mentioned, we explicitly declare that we expect a string. If we would not declare the type and someone would accidentally provide a number or a boolean, the type would automatically change, which is not what we want:

```tsx
export const BLOCK_TITLE = true;          // No error, despite the fact that we expect a string.
export const BLOCK_TITLE: string  = true; // Type 'boolean' is not assignable to type 'string'.
```

By being explicit about the type, we receive errors during development, which I believe, is what we want.

**Comment by nielslange** [comment]
> The code looks good and the block works as expected in my test. Left a minor comment about Type Inference but otherwise, this is LGTM!

Thanks for your review, @dinhtungdu. I just replied to your suggestion. Would you mind doing another review? 😀

File: assets/js/atomic/blocks/product-elements/tag-list/constants.tsx (hunk 1 above)
Comment (dinhtungdu) [reply assets/js/atomic/blocks/product-elements/tag-list/constants.tsx:14]: > By being explicit about the type, we receive errors during development, which I believe, is what we want.

Makes sense 👍🏽

**Review by dinhtungdu — APPROVED** [review]
@nielslange thanks for your clarification! Let's ship this 🚀

**Comment by nielslange** [comment]
> @nielslange thanks for your clarification! Let's ship this 🚀

Cool. Thanks, @dinhtungdu! 🙌

**Pull request merged**